weasyl/two_factor_auth.py

```diff
@@ -0,0 +1,383 @@
+    Store generated recovery codes into the recovery code table, checking for validity.
+
+    Parameters:
+        userid: The userid to save the recovery codes for.
+        recovery_codes: Comma separated unicode string of recovery codes.
+
+    Returns: Boolean True if the codes were successfully saved to the database, otherwise
+    Boolean False
+    """
+    # Force the incoming string to uppercase, then split into a list
+    codes = recovery_codes.upper().split(',')
+    # The list must exist and be equal to the current codes to generate
+    if len(codes) != _TFA_RECOVERY_CODES:
+        return False
+    # Make sure all codes are `LENGTH_RECOVERY_CODE` characters long, as expected
+    for code in codes:
+        if len(code) != LENGTH_RECOVERY_CODE:
+            return False
+
+    # Store the recovery codes securely by hashing them with bcrypt
+    hashed_codes = [bcrypt.hashpw(code.encode('utf-8'), bcrypt.gensalt(rounds=BCRYPT_WORK_FACTOR)) for code in codes]
+
+    # If above checks have passed, clear current recovery codes for `userid` and store new ones
+    d.engine.execute("""
+        BEGIN;
+
+        DELETE FROM twofa_recovery_codes
+        WHERE userid = %(userid)s;
+
+        INSERT INTO twofa_recovery_codes (userid, recovery_code_hash)
+        SELECT %(userid)s, unnest(%(recovery_code_hash)s);
+
+        COMMIT;
+    """, userid=userid, recovery_code_hash=list(hashed_codes))
+
+    # Return that we added the codes successfully
+    return True
+
+
+def is_recovery_code_valid(userid, tfa_code, consume_recovery_code=True):
+    """
+    Checks the recovery code table for a valid recovery code.
+
+    Determine if a supplied recovery code is present in the recovery code table
+    for a specified userid. If present, consume the code by deleting the record.
+    Case-insensitive, as the code is converted to upper-case before querying
+    the database.
+
+    Parameters:
+        userid: The userid of the requesting user.
+        tfa_code: A candidate recovery code to check.
+        consume_recovery_code:  If set to True, the recovery code is consumed if it is
+        valid, otherwise the call to this function is treated as a query to the validity
+        of the recovery code. (Default: Boolean True)
+
+    Returns: Boolean True if the code was valid and has been consumed, Boolean False, otherwise.
+    """
+    # Recovery codes must be LENGTH_RECOVERY_CODE characters; fast-fail if this is not the case
+    if len(tfa_code) != LENGTH_RECOVERY_CODE:
+        return False
+    # First extract the bcrypt hashes.
+    recovery_code_hash_query = d.engine.execute("""
+        SELECT recovery_code_hash
+        FROM twofa_recovery_codes
+        WHERE userid = %(userid)s
+    """, userid=userid).fetchall()
+    # Then attempt to hash the input code versus the stored code(s).
+    for row in recovery_code_hash_query:
+        if bcrypt.checkpw(tfa_code.upper().encode('utf-8'), row['recovery_code_hash'].encode('utf-8')):
```

Comment:
Just a sanity check: How much processing does this take? How feasible is it for a user to DoS us by issuing requests to recover logins with a bogus code like 11111 11111 11111 11111 repeatedly? Obviously this would only work for accounts for which they had a password.